homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,130 @@
+
+    def __init__(self, parent: "SonosEntity", soco: SoCo, battery_info: Dict[str, Any]):
+        """Initialize a SonosBatteryEntity from a parent SonosEntity."""
+        self._parent = parent
+        self._soco = soco
+        self._battery_info = battery_info
+        self._timer = None
+
+    @staticmethod
+    def fetch(soco: SoCo) -> Optional[Dict[str, Any]]:
+        """Fetch battery_info from the given SoCo object.
+
+        Returns None if the device doesn't support battery info
+        or if the device is offline.
+        """
+        try:
+            return soco.get_battery_info()
+        except (ConnectionError, TimeoutError, SoCoException):
+            pass
+        return None
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        if self._timer is None:
+            self._timer = self._parent.hass.helpers.event.async_track_time_interval(
+                self.update, SCAN_INTERVAL
+            )
+
+    # Identification of this Entity
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID of the sensor."""
+        return self._parent.unique_id + "-battery"
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._parent.name + " Battery"
+
+    @property
+    def device_info(self) -> Dict[str, Any]:
+        """Return information about the device."""
+        return self._parent.device_info
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Get the unit of measurement."""
+        return "%"
+
+    # Update the current state
+    def update(self, event=None):
+        """Poll the device for the current state."""
+        if self.available:
+            # only poll if the Sonos device is online
+            _LOGGER.debug("Starting to check battery_info on %s", self._soco)
+            battery_info = SonosBatteryEntity.fetch(self._soco)
+            if battery_info is not None:
+                self._battery_info = battery_info
+                self.schedule_update_ha_state()
+
+    # Current state
+    @property
+    def available(self) -> bool:
+        """Return whether this device is available."""
+        return self._parent.available
+
+    @property
+    def battery_level(self) -> int:
+        """Return the battery level."""
+        return self._battery_info.get("Level", 0)
+
+    @property
+    def power_source(self) -> str:
+        """Return the name of the power source.
+
+        Observed to be either BATTERY or SONOS_CHARGING_RING or USB_POWER.
+        """
+        return self._battery_info.get("PowerSource", "Unknown")
+
+    @property
+    def charging(self) -> bool:
+        """Return the charging status of this battery."""
+        return self.power_source != "BATTERY"
+
+    @property
+    def icon(self) -> str:
+        """Return the icon of the sensor."""
+        return icon_for_battery_level(self.battery_level, self.charging)
```

Comment:
If you set the device class, the frontend should take care of this for you

Reply:
It seems not to, probably because it's currently a `media_player` instead of a proper `sensor` like the refactor branch will be.

Reply:
As a sensor entity, it seems to properly show the charge level, but not the charging status.